exporter/gpu_stats.py

```diff
@@ -0,0 +1,128 @@
+from base import OSBase
+from collections import defaultdict
+from osclient import get_keystone_session, session_adapter
+from os import environ
+from prometheus_client import CollectorRegistry, generate_latest, Gauge
+import logging
+import re
+
+logger = logging.getLogger(__name__)
+
+GNOCCHI_ENDPOINT = environ['GNOCCHI_ENDPOINT']
+GRANULARITY = 60
+AGGREGATION_METHOD = 'mean'
+LABELS = ['region', 'stat_name', 'gpu_type', 'gpu_index']
+
+
+class GPUStats(OSBase):
+    """Class to report the statistics on NVIDIA GPUs"""
+
+    def __init__(self, oscache, osclient):
+        super(GPUStats, self).__init__(oscache, osclient)
+
+        self.sess = get_keystone_session()
+
+    def build_cache_data(self):
+        """Return list of stats to cache"""
+        cache_stats = []
+
+        for gpu_type, metrics in self.get_metrics_by_gpu_type().items():
+            iter = 0
+
+            for metric, gpu_indices in metrics.items():
+                metric_name = str(metric).split('.')[-1]
+
+                for gpu, metric_ids in gpu_indices.items():
+                    stat = dict(
+                        stat_name=metric_name,
+                        gpu_type=gpu_type,
+                        gpu_index=gpu)
+
+                    stat['stat_value'] = self.get_aggregate(metric_ids)
+                    cache_stats.append(stat)
+
+                    # Add gpu count if first iteration
+                    if iter == 0:
+                        count = dict(
+                            stat_name='gpu_count',
+                            gpu_type=gpu_type,
+                            gpu_index=gpu)
```

Comment:
Is there any value in having the `gpu_index` on the count metric, which I think is supposed to count up the number of GPUs? I think actually this count metric is not necessary, given that we can get it by doing a `sum(some_other_gpu_metric) by (stat_name)`

It is possible for Prometheus to define metrics that are "views" of other metrics in this manner. This helps avoid things like this, where you are creating a new metric over the raw data.

Reply:
Once we aggregate we lose any data on how many gpus are actually contributing to that aggregation. I don't think your query gets us that data back since we're exporting aggregates and not data points per node.